File: javascript/complex-numbers/complex-numbers.js
export class ComplexNumber {
    constructor(real = 0, imag = 0) {
        this.real = real;
        this.imag = imag;
    }

    add(other) {
        return new ComplexNumber(this.real + other.real, this.imag + other.imag);
    }

    sub(other) {
        return new ComplexNumber(this.real - other.real, this.imag - other.imag);
    }

    mul(other) {
        return new ComplexNumber(this.real * other.real - this.imag * other.imag,
                                 this.imag * other.real + this.real * other.imag);
    }

    div(other) {
        return new ComplexNumber((this.real * other.real + this.imag * other.imag) / (other.real**2 + other.imag**2),
                                 (this.imag * other.real - this.real * other.imag) / (other.real**2 + other.imag**2));
    }

    get abs() {
        return Math.sqrt(this.real**2 + this.imag**2);
    }

    get conj() {
        return new ComplexNumber(this.real, this.imag != 0 ? -this.imag : 0);
    }

    get exp() {
        return new ComplexNumber(Math.exp(this.real)).mul(new ComplexNumber(Math.cos(this.imag), Math.sin(this.imag)));
    }
}

File: ruby/complex-numbers/_iterations/1/complex_numbers.rb
class ComplexNumber
  DELTA = 0.01

  attr_reader :real, :imaginary

  def initialize(real, imaginary = 0)
    @real = real.to_f
    @imaginary = imaginary.to_f
  end

  def ==(other)
    (real - other.real).abs <= DELTA && (imaginary - other.imaginary) <= DELTA
  end

  def +(other)
    ComplexNumber.new(real + other.real, imaginary + other.imaginary)
  end

  def -(other)
    ComplexNumber.new(real - other.real, imaginary - other.imaginary)
  end

  def *(other)
    ComplexNumber.new(
      real * other.real - imaginary * other.imaginary,
      imaginary * other.real + real * other.imaginary
    )
  end

  def /(other)
    d = other.real ** 2 + other.imaginary ** 2
    ComplexNumber.new(
      (real * other.real + imaginary * other.imaginary) / d,
      (imaginary * other.real - real * other.imaginary) / d
    )
  end

  def abs
    Math.sqrt(real ** 2 + imaginary ** 2)
  end

  def conjugate
    ComplexNumber.new(real, -imaginary)
  end

  def exp
    ComplexNumber.new(Math::E ** real) * ComplexNumber.new(Math.cos(imaginary), Math.sin(imaginary))
  end
end
